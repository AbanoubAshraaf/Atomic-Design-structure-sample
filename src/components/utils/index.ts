export const isValidAge = (age:number): boolean => {
    return (
      age > 0 && // Age must be positive
      !isNaN(age) && // Age must be a valid number
      Number.isInteger(age) // Age must be an integer
    );
  };
  