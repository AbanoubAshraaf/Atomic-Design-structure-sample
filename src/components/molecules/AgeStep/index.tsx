import React, { useState } from "react";
import InputField from "@/components/atoms/InputField";
import Button from "@/components/atoms/Button";
import { StepProps } from "@/components/types";
import { isValidAge } from "@/components/utils";

const AgeStep: React.FC<StepProps> = ({ onClick }) => {
  const [age, setAge] = useState(0);
  const [error, setError] = React.useState<string>("");

  const validateAndProceed = () => {
    if (!isValidAge(age)) {
      setError("Please enter a valid positive integer age.");
    } else {
      setError("");
      onClick("age", age);
    }
  };

  return (
    <div className="flex flex-col gap-4 w-full">
      <InputField
        label="Age"
        value={age}
        type="number"
        error={error}
        onChange={(value) => setAge(Number(value))}
      />
      <Button text="Buy Now" onClick={validateAndProceed} />
    </div>
  );
};

export default AgeStep;
