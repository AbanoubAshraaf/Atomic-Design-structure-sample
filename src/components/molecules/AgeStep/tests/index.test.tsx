import React from "react";
import { render, fireEvent } from "@testing-library/react";
import AgeStep from "..";
import { mockValidUserData } from "@/components/test/mocks";
import { createSnapShot } from "@/components/test/utils";

describe("<AgeStep/> Component tests", () => {
  test("render age input properly", () => {
    const { container, getByTestId } = render(<AgeStep onClick={jest.fn()} />);
    expect(getByTestId("age")).toBeInTheDocument();
    createSnapShot(container);
  });

  test("call callback with age value on button click", () => {
    const mockCallback = jest.fn();
    const { getByTestId, getByText } = render(
      <AgeStep onClick={mockCallback} />,
    );

    fireEvent.change(getByTestId("age"), {
      target: { value: mockValidUserData.age },
    });
    fireEvent.click(getByText("Buy Now"));
    expect(mockCallback).toHaveBeenCalledWith("age", 20);
  });

  test("display an error for invalid decimal age", () => {
    const { container, getByTestId, getByText } = render(
      <AgeStep onClick={jest.fn()} />,
    );
    const input = getByTestId("age");

    fireEvent.change(input, { target: { value: 10.5 } });
    fireEvent.click(getByText("Buy Now"));
    expect(
      getByText("Please enter a valid positive integer age."),
    ).toBeInTheDocument();

    createSnapShot(container);
  });

  test("display an error for invalid negative age", () => {
    const { getByTestId, getByText } = render(<AgeStep onClick={jest.fn()} />);
    const input = getByTestId("age");

    fireEvent.change(input, { target: { value: -10 } });
    fireEvent.click(getByText("Buy Now"));
    expect(
      getByText("Please enter a valid positive integer age."),
    ).toBeInTheDocument();
  });

  test("does not call callback if age is invalid", () => {
    const mockCallback = jest.fn();
    const { getByTestId, getByText } = render(
      <AgeStep onClick={mockCallback} />,
    );

    fireEvent.change(getByTestId("age"), { target: { value: "0" } });
    fireEvent.click(getByText("Buy Now"));
    expect(mockCallback).not.toHaveBeenCalled();
  });
});
