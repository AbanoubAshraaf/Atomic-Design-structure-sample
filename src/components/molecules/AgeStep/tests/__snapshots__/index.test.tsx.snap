// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AgeStep/> Component tests display an error for invalid decimal age 1`] = `
<div>
  <div
    class="flex flex-col gap-4 w-full"
  >
    <div
      class="flex flex-col gap-2 items-start"
    >
      <label>
        Age
        :
      </label>
      <input
        class="input input-bordered input-primary w-full max-w-xs"
        data-testid="age"
        type="number"
        value="10.5"
      />
      <span
        class="error-text"
      >
        Please enter a valid positive integer age.
      </span>
    </div>
    <button
      class="btn btn-primary"
    >
      Buy Now
    </button>
  </div>
</div>
`;

exports[`<AgeStep/> Component tests render age input properly 1`] = `
<div>
  <div
    class="flex flex-col gap-4 w-full"
  >
    <div
      class="flex flex-col gap-2 items-start"
    >
      <label>
        Age
        :
      </label>
      <input
        class="input input-bordered input-primary w-full max-w-xs"
        data-testid="age"
        type="number"
        value="0"
      />
    </div>
    <button
      class="btn btn-primary"
    >
      Buy Now
    </button>
  </div>
</div>
`;
